Improve string types documentation and add references (#10676)
pydantic/pydantic: docs/api/standard_library_types.md

## Strings

> [`str`][]: Strings are accepted as-is.
> [`bytes`][] and [`bytearray`][] are converted using the [`decode()`][bytes.decode] method.
> Enums inheriting from [`str`][] are converted using the [`value`][enum.Enum.value] attribute.
> All other types cause an error.

Reviewer: ```suggestion
- [`str`][]: Strings are accepted as-is.
- [`bytes`][] and [`bytearray`][] are converted using the [`decode()`][bytes.decode] method.
- Enums inheriting from [`str`][] are converted using the [`value`][enum.Enum.value] attribute.

All other types cause an error.
```

## Strings

> [`str`][]: Strings are accepted as-is. [`bytes`][] and [`bytearray`][] are converted using [`v.decode()`](https://docs.python.org/3/library/stdtypes.html#bytes.decode).

Reviewer: ```suggestion
[`str`][]: Strings are accepted as-is. [`bytes`][] and [`bytearray`][] are converted using the [`decode()`][bytes.decode] method.
```

> Enums inheriting from `str` are converted using [`v.value`](https://docs.python.org/3/library/enum.html#enum.Enum.value). All other types cause an error.

Reviewer: ```suggestion
Enums inheriting from [`str`][] are converted the [`value`][enum.Enum.value] attribute. All other types cause an error.
```
